tests/unit/test_index.py

```diff
@@ -302,43 +285,116 @@ def test_get_best_candidate__best_yanked_but_not_all(self, caplog):
+        Test that the candidate_evaluator attribute is set correctly.
         """
-        search_scope = SearchScope([], [])
         selection_prefs = SelectionPreferences(
-            allow_yanked=allow_yanked,
+            allow_yanked=True,
+            allow_all_prereleases=allow_all_prereleases,
+            prefer_binary=prefer_binary,
         )
+        target_python = TargetPython(py_version_info=(3, 7, 3))
+        target_python._valid_tags = ['tag1', 'tag2']
         finder = PackageFinder.create(
-            search_scope=search_scope,
+            search_scope=SearchScope([], []),
             selection_prefs=selection_prefs,
-            session=object(),
+            session=PipSession(),
+            target_python=target_python,
         )
         evaluator = finder.candidate_evaluator
-        assert evaluator._allow_yanked == allow_yanked
+        assert evaluator.allow_all_prereleases == allow_all_prereleases
+        assert evaluator._prefer_binary == prefer_binary
+        assert evaluator._supported_tags == ['tag1', 'tag2']
 
     def test_create__target_python(self):
         """
-        Test that target_python is passed to CandidateEvaluator as is.
+        Test that the _target_python attribute is set correctly.
         """
-        search_scope = SearchScope([], [])
-        selection_prefs = SelectionPreferences(
-            allow_yanked=True,
-        )
         target_python = TargetPython(py_version_info=(3, 7, 3))
         finder = PackageFinder.create(
-            search_scope=search_scope,
-            selection_prefs=selection_prefs,
-            session=object(),
+            search_scope=SearchScope([], []),
+            selection_prefs=SelectionPreferences(allow_yanked=True),
+            session=PipSession(),
             target_python=target_python,
         )
-        evaluator = finder.candidate_evaluator
-        actual_target_python = evaluator._target_python
+        actual_target_python = finder._target_python
+        # The target_python attribute should be set as is.
         assert actual_target_python is target_python
+        # Check that the attributes weren't reset.
         assert actual_target_python.py_version_info == (3, 7, 3)
 
+    def test_create__target_python_none(self):
+        """
+        Test passing target_python=None.
+        """
+        finder = PackageFinder.create(
+            search_scope=SearchScope([], []),
+            selection_prefs=SelectionPreferences(allow_yanked=True),
+            session=PipSession(),
+            target_python=None,
+        )
+        # Spot-check the default TargetPython object.
+        actual_target_python = finder._target_python
+        assert actual_target_python._given_py_version_info is None
+        assert actual_target_python.py_version_info == CURRENT_PY_VERSION_INFO
+
+    @pytest.mark.parametrize('allow_yanked', [False, True])
+    def test_create__allow_yanked(self, allow_yanked):
```

Comment:
The test of `make_link_evaluator` is exhaustive, we can probably drop these three tests.

Reply:
On this point, these are unit tests of the `create()` method, so I think it's important to keep them. The fact that the `make_link_evaluator()` tests use the `create()` method is incidental to those tests and could change in the future. (Maybe the `make_link_evaluator()` tests should be changed to use the `__init__()` for that reason.) There's a similar reasoning with any unit tests: if an end-to-end test or any higher-level test happens to test the same code path as a unit test, we don't remove the lower-level unit tests. It's easier to know that a function is tested when we can look at tests of that function.